aten/src/ATen/native/cuda/SpectralOps.cu

```diff
@@ -0,0 +1,274 @@
+      }
+      remainder = remainder % dim_stride;
+    }
+    if (imag) {
+      return -data[read_idx + 1];
+    } else {
+      return data[read_idx];
+    }
+  }
+};
+
+// input should be a contiguous batched tensor of full signals (i.e. twosided)
+static inline void _fft_fill_with_hermitian_symmetry_(Tensor& input,
+                      int64_t size_last_dim, int64_t last_dim_start_slice) {
+  if (last_dim_start_slice >= size_last_dim) {
+    return;
+  }
+
+  // copy
+  int64_t n = input.numel() / size_last_dim * (size_last_dim - last_dim_start_slice);
+
+  cudaStream_t stream = globalContext().getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.type(), "_fft_fill_with_hermitian_symmetry_", [&] {
+    using cuda_scalar_t = cuda::type<scalar_t>;
+    typedef thrust::device_ptr<cuda_scalar_t> device_ptr;
+    typedef thrust::counting_iterator<int64_t> counter;
+    typedef thrust::transform_iterator<cnt_to_dst_idx_functor, counter> dst_idx_iterator;
+    typedef thrust::permutation_iterator<device_ptr, dst_idx_iterator> dst_iterator;
+    typedef thrust::transform_iterator<dst_idx_to_src_functor<cuda_scalar_t>, dst_idx_iterator> src_iterator;
+
+    dst_idx_iterator dst_idxs(counter(0), cnt_to_dst_idx_functor(size_last_dim, last_dim_start_slice));
+
+    auto data = device_ptr(input.data<cuda_scalar_t>());
+    dst_iterator dsts(data, dst_idxs);
+    src_iterator srcs(dst_idxs, dst_idx_to_src_functor<cuda_scalar_t>(input));
+    thrust::copy_n(policy, srcs, n, dsts);
+  });
+}
+
+// cufft
+Tensor _fft_cufft(const Tensor& self, int64_t signal_ndim,
+                  bool complex_input, bool complex_output,
+                  bool inverse, IntList checked_signal_sizes,
+                  bool normalized, bool onesided,
+                  IntList output_sizes) {
+  Tensor input = self;
+
+  bool is_half = input.type().scalarType() == ScalarType::Half;
+
+  // cuFFT requires input and output data pointers to complex type aligned
+  // our allocated output tensor is always 256 bytes aligned so it is fine, but
+  // we need to check input tensor to make sure that it is not unaligned, e.g.
+  // from a slicing.
+  auto complex_size_bytes = 2 * input.type().elementSizeInBytes();
+  if (reinterpret_cast<std::uintptr_t>(input.data_ptr()) % complex_size_bytes != 0) {
+    input = self.clone();
+  }
+
+  // check input batch size
+  long long int batch = input.size(0);
+
+  // check the input sizes and strides to see if we need to make it contiguous
+  std::vector<long long int> signal_sizes(signal_ndim);
+  std::vector<long long int> inembeds(signal_ndim);
+  long long int istride, signal_size;
+  bool need_contiguous = input.stride(signal_ndim) == 0;
+  if (complex_input) {
+    // real/imag dimension must be like complex type
+    need_contiguous |= input.stride(-1) != 1;
```

Comment:
OK, this must be the test that the real/imag dimension is done this way. It might be worth spelling this out in more words.